RosettaCommons/binder
# Implementation of cmake tests (ctest)

Hi @lyskov ,

 here is the MR we descussed in the past.

In this MR:
  - Implementation of cmake tests (ctest)
  - Usage the ctests in CI

How to use the ctest:

```
 cmake -BINDER_ENABLE_TEST=ON ....
 make  # or whatever is used 
 ctest .
```
The testing is fully enabled by default if:    cmake version is 3.0.0 or higher, pybind11 and at  python libraries/headers/interpreters are available.  Otherwise some tests will not be executed, e.g. obviously python3 tests will not be enabled if the Python3 libraries are not found.


Features:
  - The  update does not interfere with the existing python tests.

  - Multiple python versions can be and should be used in parallel. The versions are set as a comma-separated list passed by the ``BINDER_TEST_PYTHON_VERSIONS`` option. The default list is 0,2,3.
    The Python version "0" corresponds to a "plain diff" of the output vs reference. The versions of python2/Python3 will be the first     versions found by cmake.   For a specific versions one can use e.g. 
   ``   cmake  ....   -DBINDER_TEST_PYTHON_VERSIONS=0,2.7.15,3.8.0,3.7.0`` .  The generated codes will be compiled and loaded using the corresponding interpreter.

- With an option  ``-DBINDER_MOCK_TEST`` one can mock the code generation by binder. In this case the reference codes will be used in the python tests.

 - The actual comparison of the source codes is done not via GNU/Linux diff, but via simple "diffbinder" code in the repo. This assures portability (e.g. on Windows and Unixes with non-GNU diff like Solaris ) and allows to ignore certain patterns in the generated codes w/o an invocation of external programs that should treat regexps.

-  So far different versions of LLVM produce slightly different codes, so all the comparisons are run with an argument --always-success, which forces the diffbinder utility to report "success" regardless of the actual differences.  The actual difference is, of course still visible, e.g. in logs or using  ``ctest . --verbose ``


Technical things to note: 
  -  The  ``/cmake``  directory is added with some  cmake scripts from the standard cmake distribution. Some huge files, but these are standard.

Possible future extensions:
 - Include PyPy support

General observations for binder from this MR:
 - In case binder is compiled out of the LLVM source tree and the host system has installed gcc as a default compiller, one has to 
set the path to the CLang headers as an extra argument in the command line.   It would be nice to describe it in the documentation.


Tagging  @EricCousineau-TRI and link #120 

Best regards,

Andrii
